CI: Windows + CUDA

ax3l opened the pull request:
## Summary

This a CI entry for Windows MSVC + CUDA.

@robertmaynard @WeiqunZhang @mic84 

## Additional background

The install of CUDA takes about 10min, but we cache its install after the first run (this cache will be [evicted after 7 days of inactivity](https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy)).

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

ax3l committed:
CI: Windows + CUDA

Fix NVCC Hint via CMAKE_CUDA_COMPILER

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -15,3 +15,38 @@ jobs:
+        export CXX="C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe"
+        mkdir build
+        cd build
+        cmake -G "Visual Studio 16 2019" -A x64 -DAMReX_BUILD_TUTORIALS=ON -DAMReX_FORTRAN=OFF -DAMReX_MPI=OFF -DAMReX_GPU_BACKEND=CUDA -DCMAKE_CXX_STANDARD=17 -DCMAKE_CUDA_STANDARD=17 -DAMReX_CUDA_ARCH=6.0 -DCUDAToolkit_ROOT="C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.1" ..
```
@robertmaynard this still does not find the `CMAKE_CUDA_COMPILER` - should one set more/other hints on Windows?

robertmaynard replied:
The visual studio generator required the MSBuild extensions for CUDA to be installed, is that the case here? You might also need to specify the location of `CMAKE_CUDA_COMPILER`.

You might be better off trying the Ninja/Makefile generator as it doesn't require the MSBuild extensions and will have faster turnaround times.

ax3l replied:
Thanks! I think that MSBuild should be here, how can one check? What's the CMake call you used to build AMReX with CUDA?

I thought that passing the `CUDAToolkit_ROOT` variable or setting `CUDA_PATH` or setting `CUDACXX` might indicate where `nvcc` is located, is that not the hint one would pass?

robertmaynard replied:
Using CMake 3.19.1 from a normal shell the following was sufficient (For Visual Studio 2019 Generator):
```
cmake -S amrex/  -b <build_dir> \
   -DAMReX_GPU_BACKEND=CUDA \
   -DAMReX_FORTRAN=OFF \
   -DAMReX_MPI=ON \
   -DBUILD_SHARED_LIBS=OFF
```

For the Ninja generator I used:

```
cmake -S src -B nbuild -G Ninja \
   -DAMReX_GPU_BACKEND=CUDA \
   -DAMReX_FORTRAN=OFF \
   -DAMReX_MPI=ON \
   -DBUILD_SHARED_LIBS=OFF \
   -DCMAKE_CUDA_COMPILER="C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.1/bin/nvcc.exe"
```

In general if the CUDA msbuild extensions are installed, you should get nvcc detected automatically when using the Visual Studio Generator. IIRC `CUDACXX` might be ignored for the Visual Studio Generator and instead you would need to specify which CUDA version to use via `CMAKE_GENERATOR_TOOLSET` ( https://cmake.org/cmake/help/v3.19/variable/CMAKE_GENERATOR_TOOLSET.html )

ax3l replied:
Thank you, which MPI did you use on Windows? Microsoft MPI?

robertmaynard replied:
Yes, MS-MPI others should work but I didn't have those installed already.

ax3l replied:
@robertmaynard I tested all of the above mentioned options and none seems to work (see commit history of this PR)

@mic84 would work on #1639 once we get the setup here together (we have no windows at hand to test it without this CI setup)

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -15,3 +15,38 @@ jobs:
+    - name: Windows CUDA Cache
+      uses: actions/cache@v2
+      with:
+        path: "C:/Program Files/NVIDIA GPU Computing Toolkit/"
```
I hope this works. Dunno what Windows installs register these days in the system besides this.
Something, something registry? :cat2:

robertmaynard replied:
You might want the whole path to the bin directory `C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/vX.Y/bin/`

ax3l replied:
But that's more specific, no? I don't want to miss auxiliary files of the install.

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -15,3 +15,38 @@ jobs:
+        cmake -G "Visual Studio 16 2019" -A x64 -DAMReX_BUILD_TUTORIALS=ON -DAMReX_FORTRAN=OFF -DAMReX_MPI=OFF -DAMReX_GPU_BACKEND=CUDA -DCMAKE_CXX_STANDARD=17 -DCMAKE_CUDA_STANDARD=17 -DAMReX_CUDA_ARCH=6.0 -DCUDAToolkit_ROOT="C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.1" ..
+        cmake --build . --config Release
+    # use "cmd", see https://gitlab.kitware.com/cmake/cmake/-/issues/20281
+    # -DAMReX_PARTICLES=ON
```
@mic84 as soon as I activated particles, it tried to search for a Fortran compiler. Maybe something to check later on.

ax3l replied:
Maybe outdated, will check again once PR works.

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -39,3 +39,38 @@ jobs:
+              -DAMReX_FORTRAN=OFF        `
+              -DAMReX_MPI=OFF            `
+              -DAMReX_GPU_BACKEND=CUDA   `
+              -DCMAKE_CXX_STANDARD=17    `
+              -DCMAKE_CUDA_STANDARD=17   `
```
I think CUDA C++17 is generally not working atm, so let's stay with C++14 for now: #1665

ax3l committed:
CUDA/C++14

C++17 builds with NVCC are currently (up to CUDA 11.2)
failing with an internal compiler transformation bug
in extended lambda support.

ax3l committed:
Try cmd

My pwsh skills are weak:
```
CMake Error: The source directory "D:/a/amrex/amrex/.0" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
```

ax3l committed:
Skip NVCC path hint

ax3l commented:
@robertmaynard not sure if relevant to your tasks, but we enabled `.dll` support in #1847 .

With this CI build (Windows + CUDA), I am still puzzled.

sayerhs commented:
@ax3l May be add a cron option to `on:` to compile the Windows+CUDA build on a weekly basis?

robertmaynard commented:
> @robertmaynard not sure if relevant to your tasks, but we enabled `.dll` support in #1847 .
> 
> With this CI build (Windows + CUDA), I am still puzzled.

I think you want the Ninja generator correct? But the CI config doesn't have an explicit Generator specified so you are getting Visual Studio ( I think that is the problem ).

ax3l commented:
I tried `cmd` terminal and `-G Ninja` in an earlier commit (2nd last), but that did not work either.

ax3l committed:
Win CUDA CI: Ninja

ax3l committed:
CUDA: Hint CUDAToolkit_ROOT

```
CMake Error at C:/Program Files/CMake/share/cmake-3.19/Modules/CMakeDetermineCUDACompiler.cmake:166 (message):
  Could not find nvcc, please set CUDAToolkit_ROOT.
```

ax3l commented:
> May be add a cron option to on: to compile the Windows+CUDA build on a weekly basis?

I am caching the download, but if it gets to build at some point and if it still runs a long time, then we should definitely do that.

ax3l committed:
CI: Win+CUDA11 C++17 & Cleanup

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -60,3 +60,41 @@ jobs:
+        cmake -S . -B build ^
+              -G Ninja      ^
+              -DCMAKE_BUILD_TYPE=Release ^
+              -DAMReX_BUILD_TUTORIALS=ON ^
```
unused option:
```suggestion
```

ax3l committed:
remove tutorials option

ax3l commented:
I think this PR should cache it's download well after the first build.
We can re-work it in case it does not and causes overly long CI times.

ax3l commented:
re-opening to trigger CI

ax3l committed:
Windows Server 2022, CUDA 12.4

Update CI modules, OS and CUDA

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -139,3 +139,39 @@ jobs:
+        New-item -ItemType directory -Name cuda
+        Set-Location -Path cuda -PassThru
+        Invoke-WebRequest -Uri 'https://developer.download.nvidia.com/compute/cuda/12.4.0/network_installers/cuda_12.4.0_windows_network.exe' -OutFile 'cuda_install.exe'
+        Start-Process -FilePath '.\cuda_install.exe' -ArgumentList '-s' -Wait -NoNewWindow
```
```suggestion
        Start-Process -FilePath '.\cuda_install.exe' -ArgumentList '-s nvcc cudart cublas cufft curand' -Wait -NoNewWindow
        Remove-Item cuda_install.exe
```

ax3l committed:
CUDA Network Installer: Components

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -139,3 +139,40 @@ jobs:
+        if (Test-Path "C:\Program Files\NVIDIA GPU Computing Toolkit\") { exit }
+        New-item -ItemType directory -Name cuda
+        Set-Location -Path cuda -PassThru
+        Invoke-WebRequest -Uri 'https://developer.download.nvidia.com/compute/cuda/12.4.0/network_installers/cuda_12.4.0_windows_network.exe' -OutFile 'cuda_install.exe'
+        Start-Process -FilePath '.\cuda_install.exe' -ArgumentList '-s nvcc cudart cublas cufft curand' -Wait -NoNewWindow
```
Found by searching all repos on GitHub:
```suggestion
        $cuda_version_short = '12.4'
        Invoke-WebRequest -Uri 'https://developer.download.nvidia.com/compute/cuda/12.4.0/network_installers/cuda_12.4.0_windows_network.exe' -OutFile 'cuda_install.exe'

        $CudaFeatures = " nvcc_${cuda_version_short} " + `
  " cublas_${cuda_version_short} " + `
  " cublas_dev_${cuda_version_short} " + `
  " cuda_profiler_api_${cuda_version_short} " + `
  " cudart_${cuda_version_short} " + `
  " cufft_${cuda_version_short} " + `
  " cufft_dev_${cuda_version_short} " + `
  " cuobjdump_${cuda_version_short} " + `
  " cupti_${cuda_version_short} " + `
  " curand_${cuda_version_short} " + `
  " curand_dev_${cuda_version_short} " + `
  " cusolver_${cuda_version_short} " + `
  " cusolver_dev_${cuda_version_short} " + `
  " cusparse_${cuda_version_short} " + `
  " cusparse_dev_${cuda_version_short} " + `
  " cuxxfilt_${cuda_version_short} " + `
  " npp_${cuda_version_short} " + `
  " npp_dev_${cuda_version_short} " + `
  " nsight_vse_${cuda_version_short} " + `
  " nvdisasm_${cuda_version_short} " + `
  " nvjitlink_${cuda_version_short} " + `
  " nvjpeg_${cuda_version_short} " + `
  " nvjpeg_dev_${cuda_version_short} " + `
  " nvml_dev_${cuda_version_short} " + `
  " nvprof_${cuda_version_short} " + `
  " nvprune_${cuda_version_short} " + `
  " nvrtc_${cuda_version_short} " + `
  " nvrtc_dev_${cuda_version_short} " + `
  " nvtx_${cuda_version_short} " + `
  " sanitizer_${cuda_version_short} " + `
  " thrust_${cuda_version_short} "

        $SilentFlag = '-s '
        Start-Process -FilePath '.\cuda_install.exe' -ArgumentList @($SilentFlag + $CudaFeatures) -Wait -NoNewWindow
```

ax3l committed:
CUDA Features

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -139,3 +139,75 @@ jobs:
+      shell: cmd
+      run: |
+        call "C:\Program Files (x86)\Microsoft Visual Studio\2022\Enterprise\vc\Auxiliary\build\vcvarsall.bat" x64
+        SET "CUDA_PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.4"
```
For completeness we could also set `CUDA_HOME`:
```suggestion
        SET "CUDA_PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.4"
        SET "CUDA_HOME=%CUDA_PATH%"
```

ax3l committed:
Install CUDA from Script

ax3l committed:
Also set `CUDA_HOME`

ax3l committed:
Features

ax3l committed:
Fix Path

ax3l committed:
CMake: `-A x64`

ax3l committed:
CUDA Windows: Full 3GB Installer

ax3l committed:
WebInstaller, Legacy `CUDA_TOOLKIT_ROOT_DIR`

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -139,3 +139,38 @@ jobs:
+        SET "CUDA_TOOLKIT_ROOT_DIR=%CUDA_PATH%"
+        SET "CMAKE_PREFIX_PATH=%CUDA_PATH%;%CMAKE_PREFIX_PATH%"
+        SET "PATH=%CUDA_PATH%\bin\;%PATH%"
+        SET "PATH=%CUDA_PATH%\libnvvp\;%PATH%"
```
```suggestion
        SET "PATH=%CUDA_PATH%\libnvvp\;%PATH%"
        SET "CUDACXX=%CUDA_PATH%\bin\nvcc.exe"

```

ax3l committed:
Set CUDACXX

ax3l commented on .github/workflows/dependencies/windows_install_cuda.ps1:
```diff
@@ -0,0 +1,56 @@
+Start-Process -FilePath '.\cuda_install.exe' -ArgumentList @($SilentFlag + $CudaFeatures) -Wait -NoNewWindow
+
+# cleanup
+Remove-Item cuda_install.exe
```
```suggestion
Remove-Item cuda_install.exe

# VisualStudio integration
Copy-item -Force -Recurse -Verbose "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v${cuda_version_short}\extras\visual_studio_integration\MSBuildExtensions" -Destination "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\MSBuild\Microsoft\VC\v170\BuildCustomizations"
Copy-item -Force -Recurse -Verbose "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v${cuda_version_short}\extras\visual_studio_integration\MSBuildExtensions" -Destination "C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\BuildCustomizations"

# GitHub actions env updates
if (Test-Path "env:GITHUB_ACTIONS") {
    echo "Adding CUDA to CUDA_PATH and PATH"
    $env:CUDA_PATH = "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v${cuda_version_short}"
    echo "CUDA_PATH=$env:CUDA_PATH" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
    echo "$env:CUDA_PATH/bin" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
}
```

ax3l commented on .github/workflows/windows.yml:
```diff
@@ -139,3 +139,40 @@ jobs:
+
+        cmake -S . -B build ^
+              -A x64        ^
+              -DCUDAToolkit_ROOT="%CUDA_PATH%"  ^
```
```suggestion
              -DCMAKE_GENERATOR_TOOLSET="cuda=%CUDA_PATH%"  ^
              -DCUDAToolkit_ROOT="%CUDA_PATH%"  ^
```

ax3l committed:
CMAKE_GENERATOR_TOOLSET

ax3l committed:
More integrations

ax3l committed:
CMAKE_VS_PLATFORM_TOOLSET_CUDA: CUDA Version

Assume correctly installed

ax3l committed:
VS Integration Files: Auto

Should be automatically installed for VS2022:
https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/#sample-projects

WeiqunZhang commented:
I think I have found an approach that works. So I am going to close this PR.

ax3l commented:
Yay! Which approach did we do again in the end? :D

WeiqunZhang commented:
Something like
```
    steps:
    - uses: Jimver/cuda-toolkit@v0.2.19
      id: cuda-toolkit
      with:
        cuda: '12.6.1'
        sub-packages: '["nvcc", "cudart", "visual_studio_integration"]'  
        non-cuda-sub-packages: '["libcufft", "libcusparse", "libcurand"]'
        method: 'network'
        use-github-cache: 'false'
    - uses: actions/checkout@v4
    - name: Compile
      run: |
        cmake -S . -B build                       `
              -DCUDAToolkit_ROOT="$Env:CUDA_PATH" `
              -DCMAKE_VERBOSE_MAKEFILE=ON         `
              -DCMAKE_BUILD_TYPE=Release          `
              -DAMReX_GPU_BACKEND=CUDA            `
              -DAMReX_CUDA_ARCH="8.0"             `
              -DAMReX_ENABLE_TESTS=ON             `
              -DAMReX_FFT=ON                      `
              -DAMReX_FORTRAN=OFF                 `
              -DAMReX_MPI=OFF
        cmake --build build --config Release -j 4
        cmake --build build --config Release --target install
```

There are still some windows issues like not allowing extended lambda inside if constexpr and `#define small char` in some Windows system header. But it's compiling.

ax3l commented:
Great!

Another, but very Linux-y approach I was seeing the other day and try to get up for an ImpactX developer is using WSL:
https://docs.nvidia.com/cuda/wsl-user-guide/index.html

Of course, that one is not using MSVC and is essentially Ubuntu.
